<h1 align=center><font size = 6.5>Segmenting and Clustering Neighborhoods in Toronto City</font></h1>

# Introduction

In this assignment, the neighbourhoods of Toronto City will be explored. In doing so, first the neighbourhoods of Toronto city will be identified and listed. This list then will be used to explore venues in them using the Foursquare API. Once the venues for each of the neighbourhood is identified, a clustring alogorithm will be run to cluster similar neighbourhoods together. Based on the specific kind of venues available in each of the nieghbourhood they will be named. The following will define this project. 

# Table Of Content

1. Table of Postal Codes - Web Scraping & creating Dataframe.
2. Lat Long for Neighbourhood - Dataframe.
3. Plotting Neighbourhoods of Toronto on the Map. 
4. Extracting the venues of one Neighbourhood using Foursquare API.
5. Extracting the venues of all Neighbourhood using Foursquare API.
6. Exploring/Analysing Venues of all Neighbourhoods  
7. Clustering the Neighbourhoods
8. Analysisng clusters based on its properties(venues)

Before starting the assignment, the dependencies needs to be downloaded and installed. 

In [2]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


In [3]:
pip install beautifulsoup4

     |████████████████████████████████| 122kB 6.5MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [3]:
from bs4 import BeautifulSoup
import requests
import pandas as pd
import numpy as np

In [2]:
pip install lxml

     |████████████████████████████████| 5.5MB 6.0MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


# 1. Table of Postal Codes - Web Scraping & creating Dataframe

We shall be using pandas to read the wiki page link "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M" as html. 
'dfl' declares all the tables in the html page as dataframes. Amoung all the tables in the html, df is defined as the postal code table. Dataframe "df" has all the details of 'Postal Codes', 'Borough' and 'Nieghbourhoods' in Toronto city. The following data manipulation has been done to arrive at the final df table 
1.  Ensured there is no 'Not assigned' Nieghbourhood in the data frame.
2. All the Borough with Not assigned values are dropped from the dataframe. 
3. Index is reset. 


Extracting data from wiki page as a dataframe. 

In [4]:
wiki_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
dfl = pd.read_html(wiki_url)
df = dfl[0]

Ignore cells with a borough that is Not assigned.

In [5]:
df.replace('Not assigned',np.nan,inplace=True)
df.dropna(axis=0,inplace=True)

Combining Neighbourhood with same Postal Code

In [6]:
result = df.groupby(['Postal Code','Borough'], sort=False).agg( ', '.join)

In [7]:
df_new=result.reset_index()
df_new.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [8]:
df_new.shape

(103, 3)

# 2. Lat Long for Neighbourhood - Dataframe

Extracting the geographical coordinates of each postal code from the csv file @  http://cocl.us/Geospatial_data

In [9]:
df_latlon = pd.read_csv("http://cocl.us/Geospatial_data")
df_latlon.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Merging the postal code and geographical coordinates for neighbourhoods

In [27]:
canada_df = pd.merge(df_new, df_latlon, on='Postal Code')
canada_df.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

The above dataset contains the postal code, Borough, Neighbourhoods and Latitudes and longitude details of the neighbourhoods in Cananda. Since our focus area is only Toronto, we will extract all the Borough and Neighbourhood details related to the following
1. East Toronto
2. Downtown Toronto
3. West Toronto
4. Central Toronto

In [25]:
Toronto_data = canada_df[canada_df['Borough'].isin(['East Toronto', 'Downtown Toronto', 'West Toronto', 'Central Toronto'])].reset_index(drop=True)
Toronto_data.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [26]:
Toronto_data.shape

(39, 5)

This gives us the final dataset of the Neighbourhoods in Toronto city. 

# 3. Plotting Neighbourhoods of Toronto on the Map. 

The "Toronto_data" is used to superimpose the neighbourhoods on the map of Toronto city using Folium library. For this, first, the geocodes for Toronto city is identified. 

In [28]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto city are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto city are 43.6534817, -79.3839347.


Once the geocodes of Toronto City are explored, Folium library is used to superimpose the neighbourhoods on the map of Toronto city. 

In [33]:
# create map of Toronto City using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, label in zip(Toronto_data['Latitude'], Toronto_data['Longitude'], Toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

# 4. Extracting the venues of one Neighbourhood using Foursquare API.

To extract the venues of one Neighbourhood in the Toronto city, Foursquare API will be required. For this, first, the Foursquare API client credentials needs to be established. For this project, I'm using my Foursquare API credentials.

In [34]:
CLIENT_ID = 'ELAXVPSFXWALETAHQEOOH4C3KFTR0H04HURJZ4ENESZIMYXF' # your Foursquare ID
CLIENT_SECRET = 'W34MJTHDYYLQ55LYA2BCADHBOKIOPEL0HF4AXPZPZUPKFRK1' # your Foursquare Secret
VERSION = '20201104' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ELAXVPSFXWALETAHQEOOH4C3KFTR0H04HURJZ4ENESZIMYXF
CLIENT_SECRET:W34MJTHDYYLQ55LYA2BCADHBOKIOPEL0HF4AXPZPZUPKFRK1


Extracting one Neighbourhood from Toronto_data. 

In [40]:
Neigh1 = Toronto_data.loc[0, 'Neighbourhood']
Neigh1

'Regent Park, Harbourfront'

There are 2 neighbourhoods in Downtown Toronto. They are Regent Park, Harbourfront. Since both the neighbourhoods are identified under one Borough, we shall be extracting, venues from both of them. For extracting the venues we would require Geocodes which we have for Downtown Toronto. Getting the geocodes for the first 2 neighbourhoods.

In [41]:
Neigh1_latitude = Toronto_data.loc[0, 'Latitude'] # neighbourhood latitude value
Neigh1_longitude = Toronto_data.loc[0, 'Longitude'] # neighbourhood longitude value

Neigh1_name = Toronto_data.loc[0, 'Neighbourhood'] # neighbourhood name

print('Latitude and longitude values of {} are {}, {}.'.format(Neigh1_name, 
                                                               Neigh1_latitude, 
                                                               Neigh1_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


Getting the url set for Foursquare API

In [42]:
# type your answer here
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    Neigh1_latitude, 
    Neigh1_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=ELAXVPSFXWALETAHQEOOH4C3KFTR0H04HURJZ4ENESZIMYXF&client_secret=W34MJTHDYYLQ55LYA2BCADHBOKIOPEL0HF4AXPZPZUPKFRK1&v=20201104&ll=43.6542599,-79.3606359&radius=500&limit=100'

Getting the top 100 venues of the neighbourhoods with a radius of 500 mtrs. This done by "get" request. The API will return a .json file. We shall extract the venues for the first neighbourhood fron the same .json file.

In [44]:
results = requests.get(url).json()

We need to create a function that will extraxt the categories from the .jason file declared as "results". 

In [45]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

The above function is used to create a dataframe with all the list of the following
1. Venue
2. Venue Category
3. Venue location latitude
4. Venue location longitude

In [47]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3          Impact Kitchen           Restaurant  43.656369 -79.356980
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [48]:
nearby_venues.shape

(44, 4)

There are 44 venues identified for the first neighbourhood. 

# 5. Extracting the venues of all Neighbourhood using Foursquare API.

The same process will be repeated to extract venues from all the neighbourhood. For the same, the following function is required. 

In [49]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

The above function is used to create the dataframe of venues in Toronto city. 

In [50]:
Toronto_venues = getNearbyVenues(names=Toronto_data['Neighbourhood'],
                                   latitudes=Toronto_data['Latitude'],
                                   longitudes=Toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [55]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1624, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Regent Park, Harbourfront                43.65426               -79.360636   
1  Regent Park, Harbourfront                43.65426               -79.360636   
2  Regent Park, Harbourfront                43.65426               -79.360636   
3  Regent Park, Harbourfront                43.65426               -79.360636   
4  Regent Park, Harbourfront                43.65426               -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3           Restaurant  
4                  Spa

The above dataframe is the resultant dataset. It has a list of 1624 venues across Toronto city. 

In [109]:
venue_count = Toronto_venues.groupby('Neighbourhood').count()
venue_count.head()

Neighbourhood Latitude  \
Neighbourhood                                                                
Berczy Park                                                             55   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      68   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Berczy Park                                                              55   
Brockton, Parkdale Village, Exhibition Place                             23   
Business reply mail Processing Centre, South Ce...                       16   
CN Tower, King and Spadina, Railway Lands, Harb...                       16   
Central Bay Street                                                       68   

                                                    Venue  Venue Latitude  \
Neighbourhood                                                               
Berczy Park                                            55              55   
Brockton, Parkdale Village, Exhibition Place           23              23   
Business reply mail Processing Centre, South Ce...     16              16   
CN Tower, King and Spadina, Railway Lands, Harb...     16              16   
Central Bay Street                                     68              68   

                                                    Venue Longitude  \
Neighbourhood                                                         
Berczy Park                                                      55   
Brockton, Parkdale Village, Exhibition Place                     23   
Business reply mail Processing Centre, South Ce...               16   
CN Tower, King and Spadina, Railway Lands, Harb...               16   
Central Bay Street                                               68   

                                                    Venue Category  
Neighbourhood                                                       
Berczy Park                                                     55  
Brockton, Parkdale Village, Exhibition Place                    23  
Business reply mail Processing Centre, South Ce...              16  
CN Tower, King and Spadina, Railway Lands, Harb...              16  
Central Bay Street                                              68

In [58]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 236 uniques categories.


# 6. Exploring/Analysing Venues of all Neighbourhoods

To analyse the venues across all neighbourhood, a suitable dataframe needs to be created showing the presence of each category of venue in each neighbourhood. For this, the data needs to be transformed as follows. 

In [76]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighbourhood column back to dataframe
Toronto_onehot['Neighbourhood'] = Toronto_venues['Neighbourhood'] 

# move neighbourhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  \
0  Regent Park, Harbourfront                  0        0                   0   
1  Regent Park, Harbourfront                  0        0                   0   
2  Regent Park, Harbourfront                  0        0                   0   
3  Regent Park, Harbourfront                  0        0                   0   
4  Regent Park, Harbourfront                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       1     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Candy St

In [77]:
Toronto_onehot.shape

(1624, 237)

### Analysis 1
The above table gives us dataframe in which each row represents each venue in Toronto with its corrosponding Neighbourhood name. This dataset will make more sense when, the rows are grouped together along each neighbourhood taking the mean of the occarance of each category of venue in corrosponding neighbourhood. 

In [91]:
Toronto_grouped = Toronto_onehot.groupby('Neighbourhood').mean().reset_index()
Toronto_grouped.head()

Neighbourhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2  Business reply mail Processing Centre, South C...                0.0   
3  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
4                                 Central Bay Street                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0   0.0000              0.0000        0.0000           0.000            0.000   
1   0.0000              0.0000        0.0000           0.000            0.000   
2   0.0000              0.0000        0.0000           0.000            0.000   
3   0.0625              0.0625        0.0625           0.125            0.125   
4   0.0000              0.0000        0.0000           0.000            0.000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0            0.0000                  0.0           0.0       0.0     0.018182   
1            0.0000                  0.0           0.0       0.0     0.000000   
2            0.0000                  0.0           0.0       0.0     0.000000   
3            0.0625                  0.0           0.0       0.0     0.000000   
4            0.0000                  0.0           0.0       0.0     0.000000   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0    0.000000                  0.0               0.0                 0.0   
1    0.000000                  0.0               0.0                 0.0   
2    0.000000                  0.0               0.0                 0.0   
3    0.000000                  0.0               0.0                 0.0   
4    0.014706                  0.0               0.0                 0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank       Bar  \
0         0.0000        0.0         0.0    0.018182  0.036364   0.0  0.000000   
1         0.0000        0.0         0.0    0.000000  0.043478   0.0  0.043478   
2         0.0625        0.0         0.0    0.000000  0.000000   0.0  0.000000   
3         0.0000        0.0         0.0    0.000000  0.000000   0.0  0.062500   
4         0.0000        0.0         0.0    0.000000  0.000000   0.0  0.000000   

   Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  Beer Bar  \
0               0.0            0.018182  0.018182              0.0  0.036364   
1               0.0            0.000000  0.000000              0.0  0.000000   
2               0.0            0.000000  0.000000              0.0  0.000000   
3               0.0            0.000000  0.000000              0.0  0.000000   
4               0.0            0.000000  0.000000              0.0  0.000000   

   Beer Store  Belgian Restaurant    Bistro  Boat or Ferry  Bookstore  \
0         0.0                 0.0  0.018182         0.0000   0.000000   
1         0.0                 0.0  0.000000         0.0000   0.000000   
2         0.0                 0.0  0.000000         0.0000   0.000000   
3         0.0                 0.0  0.000000         0.0625   0.000000   
4         0.0                 0.0  0.000000         0.0000   0.014706   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0    0.0000                   0.0        0.018182   0.0000         0.000000   
1    0.0000                   0.0        0.086957   0.0000         0.000000   
2    0.0000                   0.0        0.000000   0.0625         0.000000   
3    0.0625                   0.0        0.000000   0.0000         0.000000   
4    0.0000                   0.0        0.000000   0.0000         0.029412   

   Building  Burger Joint  Burrito Place  Bus Line  Butcher      Café  \
0       0.0      0.000000       0.000000       0.0   0.0000  0.018182   
1       0.0      0.000000       0.043478       0.0   0.0000  0.130435   
2       0.0      0.000000       0.062500       0.0   0.0

In [90]:
Toronto_grouped.shape

(39, 237)

### Analysis 2
The above excercise can be done in a slighly different way. 
Instead of taking the mean of the occarance of category of each venue, we can group the neighbourhoods together by adding the number of each category of venue in that neighbourhood. This can help us in identifying the following 
1. How many category of venues are present in a neighbourhood.
2. Which specific category of venue is present/absent in a neighbourhood. 
3. How many number of each category of venue is present in the neighbourhood. 

eg: In the Neighbourhood "Central Bay Street", there are 4 'Cafe', 3 'Italian Restaurant' are present. 
(Please scroll towards right to find the numbers) 

In [92]:
Toronto_grouped1 = Toronto_onehot.groupby('Neighbourhood').sum().reset_index()
Toronto_grouped1.head()

Neighbourhood  Afghan Restaurant  \
0                                        Berczy Park                  0   
1       Brockton, Parkdale Village, Exhibition Place                  0   
2  Business reply mail Processing Centre, South C...                  0   
3  CN Tower, King and Spadina, Railway Lands, Har...                  0   
4                                 Central Bay Street                  0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        1                   1             1               2                2   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            1   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 1                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           1                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           1       2     0    0   
1              0          0           0           0       1     0    1   
2              1          0           0           0       0     0    0   
3              0          0           0           0       0     0    1   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   1      1                0         2   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       1              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              1          0         1   
4           0                   0       0              0          1         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               1        0                0         0   
1                     0               2        0                0         0   
2                     0               0        1                0         0   
3                     0               0        0                0         0   
4                     0               0        0                2         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     1   
1             0              1         0        0     3   
2             0              1         0        1     0   
3             0              0         0        0  

### Analysis 3 
The below set of codes and outputs are intendended to identify the top 10 occuring vennues in each Neighbourhood. 
For this first a function is developed to sort the categories of venue for each neighbourhood based on their frequency of occurance. 

In [93]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

The above function is used to create the dataframe with top 10 occurance of category of venues in each neighbourhood. 

In [94]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = Toronto_grouped['Neighbourhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...                  Park   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0    Seafood Restaurant          Cocktail Bar        Farmers Market   
1        Breakfast Spot             Nightclub           Coffee Shop   
2           Pizza Place    Light Rail Station            Skate Park   
3       Airport Service              Boutique       Harbor / Marina   
4                  Café    Italian Restaurant        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar            Restaurant           Cheese Shop   
1          Climbing Gym         Burrito Place            Restaurant   
2         Burrito Place        Farmers Market  Fast Food Restaurant   
3         Boat or Ferry   Rental Car Location                   Bar   
4           Salad Place       Bubble Tea Shop      Department Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery        Sandwich Place       Department Store  
1    Italian Restaurant          Intersection                    Bar  
2               Butcher            Restaurant       Recording Studio  
3                 Plane           Coffee Shop       Sculpture Garden  
4          Burger Joint   Japanese Restaurant        Thai Restaurant

# 5. Clustering the Neighbourhoods

In this section, the grouped data for each neighbourhood will be used to identify Neighbourhoods in Toronto city. Following lists the steps involved in Clusturing. 
1. K-means method will be used to cluster the Neighbourhoods.  
2. Clusters will be identified
3. Clusters will be plotted on the map of Toronto using Folium library. 
4. Each cluster's properties will be identified in terms of the kind of venues present in them. 

Running K-means to cluster the grouped data and identifying the clusters.  

In [95]:
# set number of clusters
kclusters = 4

Toronto_grouped_clustering = Toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

Adding the cluster details to the Toronto_sorted dataset to identify the nature of each cluster. 

In [96]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = Toronto_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

Toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               1           Coffee Shop   
1  43.662301 -79.389494               1           Coffee Shop   
2  43.657162 -79.378937               1        Clothing Store   
3  43.651494 -79.375418               1           Coffee Shop   
4  43.676357 -79.293031               1          Neighborhood   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                   Pub                Bakery   
1           Yoga Studio                 Diner            Restaurant   
2           Coffee Shop                  Café       Bubble Tea Shop   
3                  Café          Cocktail Bar            Restaurant   
4                   Pub     Health Food Store                 Trail   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Theater        Breakfast Spot                  Café   
1  Portuguese Restaurant                  Park           Music Venue   
2    Japanese Restaurant        Cosmetics Shop                 Hotel   
3               Beer Bar             Gastropub   American Restaurant   
4            Yoga Studio    Dim Sum Restaurant                 Diner   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0     Electronics Store                   Spa                 Beer Store  
1    Mexican Restaurant    Italian Restaurant                 Hobby Shop  
2             Bookstore           Pizza Place  Middle Eastern Restaurant  
3        Farmers Market                 Hotel        Japanese Restaurant  
4        Discount Store   Distribution Center           Doner Restaurant

Visualising the clusters by plotted them on the map of Toronto using Folium library.

In [98]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighbourhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# 8. Analysisng clusters based on its properties(venues)

In this section, data related to each cluster will be studied to identify its properties. There are 4 clusters created, hence there will be 4 parts of analysis. 

### Cluster 1

In [114]:
C1 = Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1,2] + list(range(5, Toronto_merged.shape[1]))]]
C1

Borough                Neighbourhood  Cluster Labels  \
29  Central Toronto  Moore Park, Summerhill East               0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
29            Playground                 Trail           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
29      Department Store           Escape Room     Electronics Store   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
29  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   10th Most Common Venue  
29       Doner Restaurant

In [104]:
C1.shape

(1, 12)

Cluster 1 spread across only 1 Borough ie "Central Toronto" and 2 neighbourhood ie "Moore Park, Summerhill East". the top categories of venues in this cluster are Playground, Trail, Yoga Studio. 
The venues in this cluster seems to be more of activity based than just resturants. 

### Cluster 2

In [113]:
C2 = Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1,2] + list(range(5, Toronto_merged.shape[1]))]]
C2

Borough                                      Neighbourhood  \
0   Downtown Toronto                          Regent Park, Harbourfront   
1   Downtown Toronto        Queen's Park, Ontario Provincial Government   
2   Downtown Toronto                           Garden District, Ryerson   
3   Downtown Toronto                                     St. James Town   
4       East Toronto                                        The Beaches   
5   Downtown Toronto                                        Berczy Park   
6   Downtown Toronto                                 Central Bay Street   
7   Downtown Toronto                                           Christie   
8   Downtown Toronto                           Richmond, Adelaide, King   
9       West Toronto                       Dufferin, Dovercourt Village   
10  Downtown Toronto  Harbourfront East, Union Station, Toronto Islands   
11      West Toronto                           Little Portugal, Trinity   
12      East Toronto                       The Danforth West, Riverdale   
13  Downtown Toronto           Toronto Dominion Centre, Design Exchange   
14      West Toronto       Brockton, Parkdale Village, Exhibition Place   
15      East Toronto                     India Bazaar, The Beaches West   
16  Downtown Toronto                     Commerce Court, Victoria Hotel   
17      East Toronto                                    Studio District   
18   Central Toronto                                      Lawrence Park   
20   Central Toronto                                   Davisville North   
22      West Toronto                      High Park, The Junction South   
23   Central Toronto                  North Toronto West, Lawrence Park   
24   Central Toronto                The Annex, North Midtown, Yorkville   
25      West Toronto                             Parkdale, Roncesvalles   
26   Central Toronto                                         Davisville   
27  Downtown Toronto                     University of Toronto, Harbord   
28      West Toronto                                 Runnymede, Swansea   
30  Downtown Toronto          Kensington Market, Chinatown, Grange Park   
31   Central Toronto  Summerhill West, Rathnelly, South Hill, Forest...   
32  Downtown Toronto  CN Tower, King and Spadina, Railway Lands, Har...   
34  Downtown Toronto                                     Stn A PO Boxes   
35  Downtown Toronto                        St. James Town, Cabbagetown   
36  Downtown Toronto             First Canadian Place, Underground city   
37  Downtown Toronto                               Church and Wellesley   
38      East Toronto  Business reply mail Processing Centre, South C...   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0                1           Coffee Shop                  Park   
1                1           Coffee Shop           Yoga Studio   
2                1        Clothing Store           Coffee Shop   
3                1           Coffee Shop                  Café   
4                1          Neighborhood                   Pub   
5                1           Coffee Shop    Seafood Restaurant   
6                1           Coffee Shop                  Café   
7                1         Grocery Store                  Café   
8                1           Coffee Shop                  Café   
9                1              Pharmacy                Bakery   
10               1           Coffee Shop              Aquarium   
11               1                   Bar           Coffee Shop   
12               1      Greek Restaurant           Coffee Shop   
13               1           Coffee Shop                 Hotel   
14               1                  Café        Breakfast Spot   
15               1                  Park                   Gym   
16               1           Coffee Shop            Restaurant   
17               1           Coffee Shop                  Café   
18               1                  Park           Swim Sc

In [116]:
print(C2.shape)
C2.Borough.unique()

(35, 13)


array(['Downtown Toronto', 'East Toronto', 'West Toronto',
       'Central Toronto'], dtype=object)

Cluster 2 is spread across 35 neighbourhoods in all 4 Borough ie 'Downtown Toronto', 'East Toronto', 'West Toronto' and 'Central Toronto'. 
All of the neighbourhoods in cluster 2, have high number of "coffee shops", "Cafe" or a pizza place.  

### Cluster 3

In [117]:
C3 = Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1,2] + list(range(5, Toronto_merged.shape[1]))]]
C3

Borough                                    Neighbourhood  \
21   Central Toronto  Forest Hill North & West, Forest Hill Road Park   
33  Downtown Toronto                                         Rosedale   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
21               2                  Park      Sushi Restaurant   
33               2                  Park                 Trail   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
21         Jewelry Store                 Trail           Escape Room   
33            Playground         Deli / Bodega     Electronics Store   

          6th Most Common Venue        7th Most Common Venue  \
21            Electronics Store  Eastern European Restaurant   
33  Eastern European Restaurant          Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21   Dumpling Restaurant            Donut Shop       Doner Restaurant  
33            Donut Shop      Doner Restaurant                Dog Run

In [118]:
print(C3.shape)
C3.Borough.unique()

(2, 13)


array(['Central Toronto', 'Downtown Toronto'], dtype=object)

Cluster 3 is spread across 3 neighbourhoods(Forest Hill North & West, Forest Hill Road Park and Rosedale) in 2 Borough ie 'Downtown Toronto' and 'Central Toronto'. 
All of the neighbourhoods in cluster 2, have high number of "Parks", "Trails" or an Electronics Store. 

### Cluster 3

In [119]:
C4 = Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1,2] + list(range(5, Toronto_merged.shape[1]))]]
C4

Borough Neighbourhood  Cluster Labels 1st Most Common Venue  \
19  Central Toronto      Roselawn               3           Music Venue   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden           Yoga Studio      Department Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
19           Escape Room     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19   Dumpling Restaurant            Donut Shop       Doner Restaurant

Cluster 4 is present in just 1 neighbourhoods ie, Roselawn in 'Central Toronto' Borough. 
Cluster 4, has high number of "Music Venue", "Gardens" and "Yoga Studios". 

In this study, 
1. All the neighbourhoods of Toronto city were identified, extracted using web scraping and listed as dataframe. 
2. This data was used to identify venues in each neighbourhoods in Toronto using the FourSquare API. 
3. Further, the venue data was structured as another dataframe and was again used to find clusters amoung the Neighbourhoods based on the category of venues in them. 
4. Clusters were identified using K-means method and were plotted on the map of toronto using Folium library. 
5. Each of the clusters were further profiled. 

This ends this study. 